subsys/bluetooth/audio/pacs.c

```diff
@@ -657,23 +699,151 @@ static sys_slist_t *pacs_get(enum bt_audio_dir dir)
+
+	/* Remove the Sink PAC and Location */
+	if (!param->snk_pac) {
+		first_to_rem = 0;
+		attrs_to_rem = PACS_SINK_PAC_CHAR_ATTR_COUNT + PACS_SINK_PAC_LOC_CHAR_ATTR_COUNT;
+	} else if (!param->snk_loc) {
+		first_to_rem = PACS_SINK_PAC_CHAR_ATTR_COUNT;
+		attrs_to_rem = PACS_SINK_PAC_LOC_CHAR_ATTR_COUNT;
+	} else {
+		first_to_rem = pacs_svc.attr_count;
+		attrs_to_rem = 0;
+	}
+
+	for (size_t i = first_to_rem + attrs_to_rem; i < pacs_svc.attr_count; i++) {
+		pacs_svc.attrs[i - attrs_to_rem] = pacs_svc.attrs[i];
+	}
+	pacs_svc.attr_count -= attrs_to_rem;
+
+	/* Set first_to_rem to the start of Source PAC Char, for cleaner offset calc */
+	first_to_rem = PACS_SINK_PAC_CHAR_ATTR_COUNT + PACS_SINK_PAC_LOC_CHAR_ATTR_COUNT;
+
+	/* Remove the Source PAC and Location */
+	if (!param->snk_pac) {
+		first_to_rem -= attrs_to_rem;
+		attrs_to_rem = PACS_SOURCE_PAC_CHAR_ATTR_COUNT +
+			       PACS_SOURCE_PAC_LOC_CHAR_ATTR_COUNT;
+	} else if (!param->snk_loc) {
+		first_to_rem = first_to_rem + PACS_SOURCE_PAC_CHAR_ATTR_COUNT - attrs_to_rem;
+		attrs_to_rem = PACS_SINK_PAC_LOC_CHAR_ATTR_COUNT;
+	} else {
+		return;
+	}
+
+	for (size_t i = first_to_rem; i < pacs_svc.attr_count; i++) {
+		pacs_svc.attrs[i - attrs_to_rem] = pacs_svc.attrs[i];
+	}
+	pacs_svc.attr_count -= attrs_to_rem;
+}
+
+int bt_pacs_register(const struct bt_bap_pacs_register_param *param)
+{
+	int err = 0;
+
+	if (atomic_test_bit(&pacs.flags, PACS_FLAG_REGISTERED)) {
+		LOG_DBG("PACS already registered");
+
+		return -EALREADY;
+	}
+
+	/* Save registration param so we can guard functions accordingly */
+	atomic_set_bit_to(&pacs.flags, PACS_FLAG_SNK_PAC, param->snk_pac);
+	atomic_set_bit_to(&pacs.flags, PACS_FLAG_SNK_LOC, param->snk_loc);
+	atomic_set_bit_to(&pacs.flags, PACS_FLAG_SRC_PAC, param->src_pac);
+	atomic_set_bit_to(&pacs.flags, PACS_FLAG_SRC_LOC, param->src_loc);
+
+	/* Remove characteristics if necessary */
+	configure_pacs_char(param);
+
+	err = bt_gatt_service_register(&pacs_svc);
+	if (err != 0) {
+		LOG_DBG("Failed to register ASCS in gatt DB");
+
+		return err;
+	}
+
+	atomic_set_bit(&pacs.flags, PACS_FLAG_REGISTERED);
+
+	return 0;
+}
+
+int bt_pacs_unregister(void)
+{
+	int err;
+	struct bt_gatt_attr _pacs_attrs[] = BT_PACS_SERVICE_DEFINITION();
+
+	if (!atomic_test_bit(&pacs.flags, PACS_FLAG_REGISTERED)) {
+		LOG_DBG("No pacs instance registered");
+		return -EALREADY;
+	}
+
+	atomic_set_bit(&pacs.flags, PACS_FLAG_UNREGISTERING);
+
+	err = bt_gatt_service_unregister(&pacs_svc);
+	/* If unregistration was succesfull, make sure to reset pacs_attrs so it can be used for
+	 * new registrations
+	 */
+	if (err != 0) {
+		LOG_DBG("Failed to unregister PACS");
+		atomic_clear_bit(&pacs.flags, PACS_FLAG_UNREGISTERING);
+
+		return err;
+	}
+
+	memcpy(&pacs_attrs, &_pacs_attrs, sizeof(struct bt_gatt_attr));
+	pacs_svc.attr_count = ARRAY_SIZE(pacs_attrs);
+
+	atomic_clear_bit(&pacs.flags, PACS_FLAG_REGISTERED);
+	atomic_clear_bit(&pacs.flags, PACS_FLAG_UNREGISTERING);
```

Comment:
Since both `bt_pacs_register` and `bt_pacs_unregister` modify the same values (`pacs_attrs`), then only 1 thread may be active in either of these functions at any one time. With the current code, Thread A can be in `bt_pacs_register` (calling `configure_pacs_char`) and Thread B in here. 

We need to guard it properly